tempesta_fw/http_sess.c

```diff
@@ -0,0 +1,731 @@
+	if (likely(next == end || *(char *)next->ptr == ';')) {
+		TFW_DBG3("%s: plain cookie value: %.*s\n", __func__,
+			 (int)chunk->len, (char *)chunk->ptr);
+		*val = *chunk;
+		return 1;
+	}
+
+	/* Add value chunks to out-string. */
+	TFW_DBG3("%s: compound cookie value found\n", __func__);
+	val->ptr = chunk;
+	TFW_STR_CHUNKN_ADD(val, 1);
+	val->len = chunk->len;
+	for (; chunk != end; ++chunk) {
+		if (*(char *)chunk->ptr == ';')
+			/* value chunks exhausted */
+			break;
+		TFW_STR_CHUNKN_ADD(val, 1);
+		val->len += chunk->len;
+	}
+	BUG_ON(TFW_STR_CHUNKN(val) < 2);
+
+	return 1;
+}
+
+/*
+ * Find Tempesta sticky cookie in an HTTP message.
+ */
+static int
+tfw_http_sticky_get(TfwHttpReq *req, TfwStr *cookie_val)
+{
+	TfwStr value = { 0 };
+	TfwStr *hdr;
+
+	/*
+	 * Find a 'Cookie:' header field in the request.
+	 * Then search for Tempesta sticky cookie within the field.
+	 * NOTE: there can be only one "Cookie:" header field.
+	 * See RFC 6265 section 5.4.
+	 * NOTE: Irrelevant here, but there can be multiple 'Set-Cookie"
+	 * header fields as an exception. See RFC 7230 section 3.2.2.
+	 * In this case, client merges them into one 'Cookie' header field
+	 * in response.
+	 */
+	hdr = &req->h_tbl->tbl[TFW_HTTP_HDR_COOKIE];
+	if (TFW_STR_EMPTY(hdr))
+		return 0;
+	tfw_http_msg_clnthdr_val(hdr, TFW_HTTP_HDR_COOKIE, &value);
+
+	return search_cookie(req->pool, &value, cookie_val);
+}
+
+#ifdef DEBUG
+#define TFW_PRINT_STICKY_COOKIE(addr, ua, sv)				\
+do {									\
+	char abuf[TFW_ADDR_STR_BUF_SIZE] = {0};				\
+	char hbuf[STICKY_KEY_MAXLEN * 2] = {0};				\
+	tfw_addr_fmt_v6(&(addr)->v6.sin6_addr, 0, abuf);		\
+	bin2hex(hbuf, tfw_sticky_key, STICKY_KEY_MAXLEN);		\
+	TFW_DBG("http_sess: calculate sticky cookie for %s,"		\
+		" ts=%#lx(now=%#lx)...\n", abuf, (sv)->ts, jiffies);	\
+	TFW_DBG("\t...secret: %.*s\n", (int)STICKY_KEY_MAXLEN * 2, hbuf);\
+	tfw_str_dprint(ua, "\t...User-Agent");				\
+} while (0)
+#else
+#define TFW_PRINT_STICKY_COOKIE(addr, ua, sv)
+#endif
+
+/*
+ * Create Tempesta sticky cookie value.
+ *
+ * Tempesta sticky cookie is based on:
+ * - HTTP request source IP address;
+ * - HTTP request User-Agent string;
+ * - Current timestamp;
+ * - The secret key;
+ */
+static int
+__sticky_calc(TfwHttpReq *req, StickyVal *sv)
+{
+	int r, addr_len;
+	TfwStr ua_value = { 0 };
+	TfwAddr *addr = &req->conn->peer->addr;
+	TfwStr *hdr, *c, *end;
+	char desc[sizeof(struct shash_desc)
+		  + crypto_shash_descsize(tfw_sticky_shash)]
+		  CRYPTO_MINALIGN_ATTR;
+	struct shash_desc *shash_desc = (struct shash_desc *)desc;
+
+	/* User-Agent header field is not mandatory and may be missing. */
+	hdr = &req->h_tbl->tbl[TFW_HTTP_HDR_USER_AGENT];
+	if (!TFW_STR_EMPTY(hdr))
+		tfw_http_msg_clnthdr_val(hdr, TFW_HTTP_HDR_USER_AGENT,
+					 &ua_value);
+
+	addr_len = tfw_addr_sa_len(addr);
+
+	memset(desc, 0, sizeof(desc));
+	shash_desc->tfm = tfw_sticky_shash;
+	shash_desc->flags = 0;
+
+	TFW_PRINT_STICKY_COOKIE(addr, &ua_value, sv);
```

Comment:
Maybe `_DBG_PRINT_` or `_DPRINT_`? I didn't get it at first, so I may not be alone.